### PR TITLE
Fix: karmada-controller-manager: panic: runtime error: index out of range

### LINKED ISSUE #3854: karmada-controller-manager: panic: runtime error: index out of range
**What happened**:

`karmada-controller-manager` crashed, caused by `panic: runtime error: index out of range`

**What you expected to happen**:

index checked, not to panic, not to crash.

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

The `index` may be `-1`.
https://github.com/karmada-io/karmada/blob/2526de856ebe8b4e57cdf5ad9e544604aa9f0e43/pkg/modeling/modeling.go#L162-L168

https://github.com/karmada-io/karmada/blob/2526de856ebe8b4e57cdf5ad9e544604aa9f0e43/pkg/modeling/modeling.go#L105-L114

**Environment**:
- Karmada version: v1.6.1
- kubectl-karmada or karmadactl version (the result of `kubectl-karmada version` or `karmadactl version`):
- Others:

### PR DESCRIPTION
**Pull request by halfrost**
**What type of PR is this?**

**What this PR does / why we need it**:
Fix issue #3854 
karmada-controller-manager: panic: runtime error: index out of range

**Which issue(s) this PR fixes**:
Fixes #3854 

**Special notes for your reviewer**:
none

### CODE REVIEW

**Comment by RainbowMango** [comment]
Thanks @halfrost, appreciate it!

cc @tedli @liangyuanpeng @NickYadance @jwcesign @chaosi-zju help to take a look~ Thanks.

**Comment by jwcesign** [comment]
/lgtm

These two functions are not used:
https://github.com/karmada-io/karmada/blob/920f5bf33e8df5c092e8e12202c697a0ae85c217/pkg/modeling/modeling.go#L328
https://github.com/karmada-io/karmada/blob/920f5bf33e8df5c092e8e12202c697a0ae85c217/pkg/modeling/modeling.go#L274


Can we delete them? @halfrost

**Comment by halfrost** [comment]
@jwcesign Hi, I have already removed these two functions that are currently not in use. This delete function took me some time to debug when it was initially developed. Now, they are no longer needed. Could you please review it again?

**Comment by jwcesign** [comment]
Can you help squash the multiple commits into one? other lgtm

cc @chaosi-zju for take another look

**Comment by halfrost** [comment]
@jwcesign Done

File: pkg/modeling/modeling.go [hunk 1]
```diff
@@ -16,12 +16,14 @@ import (
...
 var (
 	mu                  sync.Mutex
-	modelSortings       [][]resource.Quantity
 	defaultModelSorting []clusterapis.ResourceName
 )
 
 // ResourceSummary records the list of resourceModels
-type ResourceSummary []resourceModels
+type ResourceSummary struct {
+	Rms           []resourceModels
```
Comment (chaosi-zju) [inline pkg/modeling/modeling.go:24]: I think it is better to rename as `RMs` to represent `ResourceModels`, for `Rms` confused me at a galance.

**Review by chaosi-zju — COMMENTED** [review]
generally looks good to me

File: pkg/modeling/modeling.go (hunk 1 above)
Comment (halfrost) [reply pkg/modeling/modeling.go]: Done

File: pkg/controllers/status/cluster_status_controller.go [hunk 2]
```diff
@@ -618,8 +618,8 @@ func getAllocatableModelings(cluster *clusterv1alpha1.Cluster, nodes []*corev1.N
 		modelingSummary.AddToResourceSummary(modeling.NewClusterResourceNode(nodeAvailable))
 	}
 
-	m := make([]clusterv1alpha1.AllocatableModeling, len(modelingSummary))
-	for index, resourceModel := range modelingSummary {
+	m := make([]clusterv1alpha1.AllocatableModeling, len(modelingSummary.Rms))
+	for index, resourceModel := range modelingSummary.Rms {
```
Comment (chaosi-zju) [inline pkg/controllers/status/cluster_status_controller.go:622]: may you missed this modification ?

Comment (halfrost) [reply pkg/controllers/status/cluster_status_controller.go]: Done

**Review by RainbowMango — APPROVED** [review]
/approve

**Comment by liangyuanpeng** [comment]
Sorry for my late and  I have tried triggering ResourceSummary.getIndex continuously in an environment with 15 members of the cluster, but cannot reproduce the problem. I believe this PR can fix the problem.
/lgtm

**Comment by tedli** [comment]
Hi, @halfrost 

It's this issue fixed by a single commit which asscoiated with this pr? Is there any other commits related?

I didn't use the latest release, I have a my own branch, I cherry-pick this commit, it builds, run, and works.

However, I run into another index out of range panic, so I thought whether my cherry pick lack some commits.

```
I1207 10:05:29.583572       7 controller.go:118] "Observed a panic in reconciler: runtime error: index out of range [2] with length 2" controller="cluster" controllerGroup="cluster.karmada.io" controllerKind="Cluster" Cluster="db-shbt" namespace="" name="db-shbt" reconcileID=5fff6af8-e7dc-4716-a6cf-34c4e9a6a0e5
panic: runtime error: index out of range [2] with length 2 [recovered]
	panic: runtime error: index out of range [2] with length 2

goroutine 1559 [running]:
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile.func1()
	C:/Users/lizhen/Documents/Go/src/github.com/karmada-io/karmada/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:119 +0x1e5
panic({0x22b3fc0?, 0xc01306bc80?})
	C:/Program Files/Go/src/runtime/panic.go:914 +0x21f
github.com/karmada-io/karmada/pkg/modeling.(*ResourceSummary).getIndex(0xc00e7730a8, {0x2?, 0xc02121cf60?})
	C:/Users/lizhen/Documents/Go/src/github.com/karmada-io/karmada/pkg/modeling/modeling.go:125 +0x185
github.com/karmada-io/karmada/pkg/modeling.(*ResourceSummary).AddToResourceSummary(0xc00e7730a8, {0x1, 0xc02121cf60})
	C:/Users/lizhen/Documents/Go/src/github.com/karmada-io/karmada/pkg/modeling/modeling.go:181 +0x37
github.com/karmada-io/karmada/pkg/controllers/status.getAllocatableModelings(0xc02609e000, {0xc025e15800, 0x84, 0xc025e6d000?}, {0xc025e6d000, 0x532, 0x9f877eb7cc?})
```

**Comment by tedli** [comment]
> Hi @tedli Here is another PR: #4176

Thx :smiling_face_with_three_hearts:

**Pull request merged**